Jare style

jaremieromer opened the pull request:
Lots to unpack here:

Major work done on artistic style shader for the game.

1. closes #805, closes #804 
2.  Refactored lighting algorithms for point, spot and directional lights.
3. Added option to allow hatching texture to materials
4. Added a Noise pass that applies noise to the scene. This noise can be dampened by a gradient, and is dampened by luminosity.
5. Alphabetized the sample manifest and `Prefabs.h/cpp` so I can find things easier.
6. Fixed the bug with the depth test for Particles I introduced when I refactored `PassBackend`
7. Moved the `PostProcessPropertyBuffer` to `PerPassResourceSignature` and changed the underlying buffer type to dynamic buffer. I also added a new buffer for the noise pass properties.
8. Added additional properties to MaterialData and the Post Process passes, and exposed editor UI for them.

jaremieromer committed:
Added guy, tweaked outline width

jaremieromer committed:
Merge branch 'vnext' of https://github.com/NcStudios/NcEngine into jare-style

jaremieromer committed:
Fixed texture banding, experimenting with effects

jaremieromer committed:
Good progress

jaremieromer committed:
Happy, need cleaning up

jaremieromer committed:
Removed gradient

jaremieromer committed:
Merge branch 'vnext' of https://github.com/NcStudios/NcEngine into jare-style

jaremieromer committed:
Added mipmapping support

jaremieromer committed:
Experimenting with comic pens

jaremieromer committed:
Merge branch 'jare-mipmaps' of https://github.com/NcStudios/NcEngine into jare-style

jaremieromer committed:
Second normals pass

jaremieromer committed:
Modulating by depth

jaremieromer committed:
Properly allow parameter to control details level

jaremieromer committed:
Lighting updated, bugs fixed

jaremieromer committed:
Bazinga

jaremieromer committed:
Need a tremendous amount of cleanup

jaremieromer committed:
Need to add UI option for gradient buffer

jaremieromer committed:
Just need to set defaults

jaremieromer committed:
Object based gradient

jaremieromer committed:
Defaults established

jaremieromer committed:
Ready for cleanup

jaremieromer committed:
Merge branch 'vnext' of https://github.com/NcStudios/NcEngine into jare-style

jaremieromer committed:
More cleanup

jaremieromer committed:
Cleanup and tests passing

jaremieromer committed:
Cleanup

jaremieromer committed:
Reorder error fix

jaremieromer committed:
Cleanup of Rename All affecting old shader

McCallisterRomer commented on include/ncengine/graphics/PostProcess.h:
```diff
@@ -19,6 +20,7 @@ struct PostProcessPassFlag
     static constexpr auto None    = type{0};
     static constexpr auto Outline = type{1 << 0};
     static constexpr auto Fxaa    = type{2 << 0};
+    static constexpr auto Noise   = type{3 << 0};
```
That's not a flag bit, use `1 << 2`. Fxaa, while technically correct, should be changed too.

McCallisterRomer commented on sample/source/scenes/PhysicsTest.cpp:
```diff
@@ -1296,13 +1298,18 @@ void PhysicsTest::Load(ecs::Ecs world, ModuleProvider modules)
 void PhysicsTest::Unload()
 {
-    m_sampleUI->SetWidgetCallback(nullptr);
+    if (m_sampleUI)
```
Nobody is able to null this pointer.

McCallisterRomer commented on source/ncengine/graphics2/PostProcess.cpp:
```diff
@@ -8,7 +8,7 @@ namespace nc
 auto PassHasProperties(PostProcessPassFlag::type pass) -> bool
 {
-    return pass == PostProcessPassFlag::Outline;
+    return pass == PostProcessPassFlag::Outline || pass == PostProcessPassFlag::Noise;
```
We should be able to check against a mask for these, but the pass id bug I commented about will cause it to fail, depending on the hash result:
```cpp
 constexpr auto mask = PostProcessPassFlag::Outline | PostProcessPassFlag::Noise;
 return pass & mask;
```

jaremieromer replied:
Done

McCallisterRomer commented on source/ncengine/graphics2/ShaderTypes.h:
```diff
@@ -30,9 +30,21 @@ struct OutlinePassData
     Vector3 color = Vector3::Zero();
     float width = 1.0f;
     float depthThreshold = 0.8f;
+    float viewDirDepthThreshold = 0.4f;
```
Was done previously (quite possibly by me), but the buffer for these is created based on `OutlinePassData` but is copied directly from `OutlinePassProperties`. They're identical, so its fine ATM, but easy to get out of sync. I feel like we can just use `using OutlinePassData = OutlinePassProperties` here to avoid having to do like an identity transformation between types.

McCallisterRomer commented on source/ncengine/graphics2/ShaderTypes.h:
```diff
@@ -30,9 +30,21 @@ struct OutlinePassData
 };
 
+// Object model for noise pass properties used by post processing effects (type: constant buffer)
+struct NoisePassData
```
If we change `NoisePassProperties::noiseTex` from a `TextureView` to just the texture index, we can make this a using declaration too.

jaremieromer replied:
Not doing, based on our conversation today. (We don't want to have an integer counter in the UI, but rather the texture interface.)

McCallisterRomer commented on source/ncengine/graphics2/diligent/resource/PostProcessPropertyBufferResource.cpp:
```diff
@@ -6,33 +6,58 @@
 
+namespace
+{
+auto ToNoisePassData(const nc::NoisePassProperties& properties) -> nc::graphics::NoisePassData
```
Not needed, pending a previous comment.

jaremieromer replied:
Leaving it in - per conversation today reminding about needing the Texture View interface in the UI.

McCallisterRomer commented on source/ncengine/graphics2/frontend/subsystem/TransformCache.cpp:
```diff
@@ -64,6 +65,7 @@ void TransformCache::UpdateMatrices(ecs::ExplicitEcs<Transform> ecs)
 
         const auto handle = static_cast<TransformDataHandle>(i);
         data[handle].modelMatrix = pool.Get(entity).TransformationMatrix();
+        data[handle].invModelMatrix = DirectX::XMMatrixTranspose(DirectX::XMMatrixInverse(nullptr, data[handle].modelMatrix));
```
br00tal. This does result in a noticeable drop in performance. I don't think it will matter for the game though b/c we can pack all the statics up front.

jaremieromer replied:
Removed

McCallisterRomer commented:
The shader looks really cool!

McCallisterRomer commented:
There's a bug in hiding with `PassManifest`/pass ids. I don't think it was introduced here, and I doubt it causing any problems with the current state of things. The member `PassDesc::id` in the final `PostProcessPass` is calculated based on a hash. All of the other pass ids are based on flags. This will cause any flag-based decisioning applied to the final pass to act on it randomly.

Also `PassManifest::m_ids` is referenced searched in `PassManifest::RegisterPass()`, but nothing ever sets those ids.

jaremieromer commented:
> There's a bug in hiding with `PassManifest`/pass ids. I don't think it was introduced here, and I doubt it causing any problems with the current state of things. The member `PassDesc::id` in the final `PostProcessPass` is calculated based on a hash. All of the other pass ids are based on flags. This will cause any flag-based decisioning applied to the final pass to act on it randomly.
> 
> Also `PassManifest::m_ids` is referenced searched in `PassManifest::RegisterPass()`, but nothing ever sets those ids.

Fixed. `m_finalPass` is never added to m_postProcessPasses, so we never do any checks against the pass flags during rendering. The only place we used the flag was `MakePostProcessPassInstances`, which I've fixed by circumventing that method for `m_finalPass`. Now, `m_finalPass` is completely separate from flags.

Great catch

jaremieromer committed:
Code review feedback

McCallisterRomer commented on include/ncengine/graphics/PostProcess.h:
```diff
@@ -17,10 +17,10 @@ struct PostProcessPassFlag
-    static constexpr auto Outline = type{1 << 0};
-    static constexpr auto Fxaa    = type{2 << 0};
-    static constexpr auto Noise   = type{3 << 0};
+    static constexpr auto None    = type{1 << 0};
```
Decrement each of these - e.g. None should be 0

McCallisterRomer commented on source/ncengine/graphics2/NcGraphicsImpl2.cpp:
```diff
@@ -162,7 +162,7 @@ NcGraphicsImpl2::NcGraphicsImpl2(const config::GraphicsSettings& graphicsSetting
             {
                 PassDesc{
-                    .id = MaterialPassFlag::Depth,
+                    .flag = MaterialPassFlag::Depth,
```
sick

jaremieromer committed:
Fixed math
